src/cluster_migrateslots.c

```diff
@@ -0,0 +1,2118 @@
+static void updateSlotMigrationJobState(slotMigrationJob *job,
+                                        slotMigrationJobState state);
+static void sendSyncSlotsMessage(slotMigrationJob *job, const char *subcommand);
+static void proceedWithSlotMigration(slotMigrationJob *job);
+static slotMigrationJob *createSlotExportJob(clusterNode *target_node,
+                                             list *slot_ranges);
+static bool isSlotExportPauseTimedOut(slotMigrationJob *job);
+static void resetSlotMigrationJob(slotMigrationJob *job);
+static void finishSlotMigrationJob(slotMigrationJob *job,
+                                   slotMigrationJobState state,
+                                   char *message);
+static void freeSlotMigrationJob(void *o);
+static sds generateSlotMigrationJobDescription(slotMigrationJob *job,
+                                               clusterNode *node);
+static void slotExportTryUnpause(void);
+
+/* Create an empty list of slot ranges. */
+list *createSlotRangeList(void) {
+    list *slot_ranges = listCreate();
+    listSetFreeMethod(slot_ranges, zfree);
+    return slot_ranges;
+}
+
+/* Represent the provided slot range list as a string. */
+sds representSlotRangeList(list *slot_ranges) {
+    sds res = sdsempty();
+    listNode *ln;
+    listIter li;
+    int first = 1;
+    listRewind(slot_ranges, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        if (first) {
+            res = sdscatfmt(res, "%u-%u", range->start_slot, range->end_slot);
+            first = 0;
+        } else {
+            res = sdscatfmt(res, " %u-%u", range->start_slot, range->end_slot);
+        }
+    }
+    return res;
+}
+
+/* Return the node that owns the provided slot ranges, or NULL if there is no
+ * such node. */
+clusterNode *getClusterNodeBySlotRanges(list *slot_ranges, int *cross_node) {
+    clusterNode *n = NULL;
+    listNode *ln;
+    listIter li;
+    listRewind(slot_ranges, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        for (int i = range->start_slot; i <= range->end_slot; i++) {
+            if (server.cluster->slots[i] == NULL) {
+                return NULL;
+            }
+            if (!n) {
+                n = server.cluster->slots[i];
+            }
+            if (n != server.cluster->slots[i]) {
+                if (cross_node) *cross_node = 1;
+                return NULL;
+            }
+        }
+    }
+    return n;
+}
+
+/* Return whether or not a given slot is in the list of slot ranges. */
+bool isSlotInSlotRanges(int slot, list *slot_ranges) {
+    /* Loop to check if the slot in any slot range. */
+    listNode *ln;
+    listIter li;
+    listRewind(slot_ranges, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        if (slot >= range->start_slot && slot <= range->end_slot) {
+            return 1;
```

Comment:
nit - the more idiomatic way 
```suggestion
            return true;
```

Reply:
Right, missed this when moving from int